jet_tracking/testscreen.py

```diff
@@ -0,0 +1,170 @@
+    self.stopper = stopper
+    self.pulse_picker = pulse_picker
+    self.wave8 = wave8
+    self.params = params '''
+
+    # devices are not connected, use 'fake' PVs instead
+    self.jt_input = jt_input
+    self.jt_output = jt_output
+    self.jt_fake = jt_fake
+
+  def run(self):
+    while not self.isInterruptionRequested():
+      # check if stopper is in
+      ''' if (jt_utils.get_stopper(self.stopper) == 1): '''
+      if (self.jt_fake.stopper.get() == 1):
+        # if stopper is in, stop jet tracking
+        print('Stopper in - TRACKING STOPPED')
+        self.requestInterruption()
+        continue
+    
+      # check if pulse picker is closed
+      ''' if (jt_utils.get_pulse_picker(self.pulse_picker) == 1): '''
+      if (self.jt_fake.pulse_picker.get() == 1):
+        # if pulse picker is closed, stop jet tracking
+        print('Pulse picker closed - TRACKING STOPPED')
+        self.requestInterruption()
+        continue
+
+      # check wave8
+      ''' if (jt_utils.get_wave8(self.wave8) < self.params.thresh_w8):
+        if wave8 is below threshold, continue running jet tracking but do not move
+        print('Wave8 below threshold - NOT TRACKING')
+        sleep(2)
+        continue '''
+
+      # OR check number of frames passed? used during testing w/ 'fake' PVs
+      # get nframes timestamp
+      t_nframe = self.jt_output.nframe.get()[1] 
+      if (self.jt_output.nframe.get()[0] < self.jt_input.nframe.get() * 0.8):
+        # if too few frames passed, continue running jet tracking but do not move
+        print('Too few frames passed - NOT TRACKING')
+        sleep(2)
+        continue
+
+      # check CSPAD
+      # make sure nframes (or wave8, if that is what is used) timestamp matches with CSPAD timestamp
+      if (t_nframe != self.jt_output.det.get()[1]):
+        # if the timetamps do not match try again
+        continue
+
+      ''' if (jt_utils.get_cspad(self.cspad) < self.params.thresh_lo): '''
+      # params IOC is down, hardcode threshold 
+      if (self.jt_output.det.get()[0] < 0.45):
+        # if CSPAD is below lower threshold, move jet
+        ''' if (not self.params.bypass_camera()): '''
+        # params IOC is down, hardcode bypass
+        if (not False):
+          # if camera is not bypassed, check if there is a jet and location of jet
+          try:
+            '''jet_control._jet_calculate_step(self.camera, self.params)
+            # if jet is more than certain microns away from x-rays, move jet using camera feedback
+            if (self.params.jet_x.get() > self.params.thresh_cam):
+              jet_control._jet_move_step(self.injector, self.camera, self.params)
+              sleep(1) # change to however long it takes for jet to move
+              continue '''
+
+            # devices are not connected, print status message instead
+            print('Detector below lower threshold - MOVING JET')
+            sleep(1) # change to however long it takes for jet to move
```

Comment:
I see this often, is there a way to utilize epics callbacks to avoid this.  Maybe that's the plan.

Reply:
This is only a test case anyway, I think it's not in actual motion control, I see ophyd wraps their API around callbacks/waits